src/main/java/org/weakref/jmx/ManagedClass.java

```diff
@@ -0,0 +1,147 @@
+import com.google.common.collect.ImmutableSet;
+
+import java.lang.ref.WeakReference;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+import static org.weakref.jmx.ReflectionUtils.getAttributeName;
+import static org.weakref.jmx.ReflectionUtils.isGetter;
+
+public class ManagedClass
+{
+    private final WeakReference<Object> target;
+    private final Map<String, ManagedClass> children;
+    private final Map<String, ManagedAttribute> attributes;
+
+    private ManagedClass(Object target, Map<String, ManagedClass> children, Map<String, ManagedAttribute> attributes)
+    {
+        this.target = new WeakReference<>(target);
+        this.children = children;
+        this.attributes = attributes;
+    }
+
+    public static ManagedClass fromExportedObject(Object target)
+    {
+        Preconditions.checkNotNull(target, "target is null");
+
+        ImmutableMap.Builder<String, ManagedAttribute> attributes = ImmutableMap.builder();
+        ImmutableMap.Builder<String, ManagedClass> children = ImmutableMap.builder();
+
+        for (Map.Entry<Method, Method> entry : AnnotationUtils.findManagedMethods(target.getClass()).entrySet()) {
+            Method concreteMethod = entry.getKey();
+            Method annotatedMethod = entry.getValue();
+
+            if (isGetter(concreteMethod)) { // is it an attribute?
+
+                String attributeName = AnnotationUtils.getName(annotatedMethod);
+                String description = AnnotationUtils.getDescription(annotatedMethod);
+
+                if (attributeName == null || attributeName.equals("")) {
+                    attributeName = getAttributeName(concreteMethod);
+                }
+
+                if (AnnotationUtils.isNested(annotatedMethod) || AnnotationUtils.isFlatten(annotatedMethod)) { // is this nested?
+                    try {
+                        Object childTarget = concreteMethod.invoke(target);
+                        if (childTarget != null) {
+                            children.put(attributeName, fromExportedObject(childTarget));
+                        }
+                    }
+                    catch (ReflectiveOperationException e) {
+                        // Ignore and continue
+                    }
+                }
+
+                attributes.put(attributeName, new ManagedAttribute(concreteMethod, attributeName, description, AnnotationUtils.isFlatten(concreteMethod)));
+            }
+        }
+        return new ManagedClass(target, children.build(), attributes.build());
+    }
+
+    public boolean isTargetPresent()
+    {
+        return target.get() != null;
+    }
+
+    public Class getTargetClass()
+    {
+        if (target.get() == null) {
+            throw new IllegalStateException("Reference to target is no longer present.");
+        }
+        return target.get().getClass();
```

Comment:
The reference can disappear in between the `if` and this line. You need to get a strong reference to the object at the beginning and use that instead:

```java
Object reference = target.get();
if (reference == null) {
    throw ...;
}
return reference.getClass();
```

Reply:
I refactored getTarget() to do this and reused it in all the other places.